## Problem Statement

### Business Context

The healthcare industry is evolving quickly, and professionals must manage large amounts of medical data while making fast, accurate decisions. Information overload makes it difficult to identify the most relevant research, especially during emergencies when time is critical. Access to trusted, up-to-date medical knowledge is essential for accurate diagnoses and effective treatment plans.

Healthcare centers can improve care and efficiency by integrating systems that centralize medical information, support rapid decision-making, and give providers continuous access to reliable resources. These tools help streamline workflows, reduce errors, and enhance patient outcomes.

**Common Questions to Answer**

1. What is the protocol for managing sepsis in a critical care unit?

2. What are the common symptoms of appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

3. What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

4. What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

5. What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

### Objective

The task is to develop a RAG-based AI solution that uses trusted medical manuals to tackle key healthcare challenges. The goal is to understand issues like information overload, apply AI to streamline decision-making, analyze its impact on diagnostics and patient outcomes, evaluate its potential to standardize care, and create a functional prototype that demonstrates its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [1]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 51.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 230.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 137.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 259.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 242.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.19.0 requires numpy<2.2.0,>=1.26.0, but you have numpy 2.3.5 which is incompatible.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.3.5 which is incompatible.
opencv-python-headless 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.

In [2]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.3 pandas==2.2.2 tiktoken==0.12.0 pymupdf==1.26.5 langchain==0.3.27 langchain-community==0.3.31 chromadb==1.1.1 sentence-transformers==5.1.1 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 3.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 6.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.3/564.3 kB 23.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 60.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 55.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 87.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 93.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 26.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 101.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 26.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 52.1 MB/s eta 0

In [1]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader, PyPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## Question Answering using LLM

#### Downloading and Loading the model

In [2]:
# Defining the Hugging Face model repository to download from
# This specifies which pre-trained model to use from TheBloke's collection
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"

# Specifying the exact GGUF quantization file to load
# The ".Q6_K" indicates the quantization level (6-bit K-quants - good balance of quality/size)
# GGUF is the file format used by llama.cpp and compatible loaders
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"

In [3]:
# Downloading the specified GGUF model file from Hugging Face to local cache
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [4]:
# Initializing the Llama model with GPU offloading and batch processing
llm = Llama(
    model_path=model_path,
    n_ctx=2300,
    n_gpu_layers=38,
    n_batch=512
)

AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [5]:
# Generating a text completion from the LLM with specified decoding parameters
def response(query,max_tokens=512,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [49]:
# Assigning query to a variable
query1 = "What is the protocol for managing sepsis in a critical care unit?"
# Printing the response
print(response(query1))

Llama.generate: prefix-match hit




Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:

1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.
2. Resuscitation: Provide adequate fluid resuscitation to maintain adequate tissue perfusion. The goal is to achieve a mean arterial pressure (MAP) of at least 65 mmHg and a central venous oxygen saturation (ScvO2) of greater than 70%.
3. Antibiotics: Administer broad-spectrum antibiotics as soon as possible based on the suspected source of infection and local microbiology data.
4. Source control: Identify and address the source of infection, if possible. This may 

The query was about sepsis, and the model provided a correct and detailed response explaining the general management steps in a critical care unit.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [50]:
# Assigning query to a variable
query2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
# Printing the response
print(response(query2))

Llama.generate: prefix-match hit




Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain may start as a mild discomfort around the navel or in the lower right abdomen, which then gradually moves to the right lower quadrant and becomes more severe over time. The pain may be constant or intermittent and is often worsened by movement, coughing, or deep breathing.
2. Loss of appetite: People with appendicitis may lose their appetite due to abdominal pain and discomfort.
3. Nausea and vomiting: Vomiting is a common symptom of appendicitis, especially in the later stages of the condition.
4. Fever: A fever of 100.4°F (38°C) or higher may be present in some cases of appendicitis.
5. Constipation or diarrhea: Both constipation and diarrhea can occur with appendicitis, depending on the location and severity

The query contains three questions-
1. Symptoms of Appendicitis
2. Can be cured via medicine or not
3. Surgical Procedure Recommendation

Model provided correct answers to all the three questions with detailed description.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [51]:
# Assigning query to a variable
query3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
# Printing the response
response(query3)

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.\n\nThere are several treatments that have been shown to be effective in addressing sudden patchy hair loss:\n\n1. Corticosteroids: These are anti-inflammatory drugs that can help reduce inflammation and suppress the immune system's attack on the hair follicles. They can be applied topically or taken orally, depending on the severity of the condition.\n2. Minoxidil: This is a medication that has been shown to promote hair growth in some people with alopecia areata. It works by increa

The query contains two questions — the treatment for patchy hair loss and the cause of patchy hair loss. The model provided correct and detailed answers to both questions, including appropriate medical terminology and treatment names.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [52]:
# Assigning query to a variable
query4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
# Printing the response
response(query4)

Llama.generate: prefix-match hit


"\n\nA person who has sustained a physical injury to brain tissue, also known as a traumatic brain injury (TBI), may require various treatments depending on the severity and location of the injury. Here are some common treatments recommended for TBIs:\n\n1. Emergency care: The first priority is to ensure the person's airway is clear, they are breathing, and their heart is beating normally. In severe cases, emergency surgery may be required to remove hematomas or other obstructions.\n2. Medications: Depending on the symptoms, medications may be prescribed to manage conditions such as swelling, seizures, pain, or infections. For example, corticosteroids may be used to reduce brain swelling, and anticonvulsants may be given to prevent seizures.\n3. Rehabilitation: Rehabilitation is an essential part of the recovery process for TBI patients. Rehabilitation may include physical therapy to help with mobility and strength, occupational therapy to help with daily living skills, speech therapy 

The query is asking about the treatment for physical injury to brain tissue for which model has given accurate descriptive response.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [53]:
# Assigning query to a variable
query5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# Printing the response
response(query5)

Llama.generate: prefix-match hit


"\n\nFirst and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:\n\n1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.\n2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.\n3. Immobilize the leg: Use a splint, sling, or other available materials to immobilize the leg and prevent movement. Be sure not to apply too much pressure on the injury site.\n4. Provide pain relief: Offer over-the-counter pain medication, such as acetaminophen or ibuprofen, to help manage pain.\n5. Seek medical attention: If the fracture is severe or if you suspect that there may be other injuries, seek medical help as soon as possible.\n\nOnce you've ensured the person's safety and stabi

The query contains two questions—treatment steps for a fractured leg and the care and recovery process. The model provided accurate and complete answers to both, with an appropriate emphasis on the recovery and care aspects, which is expected.

## Question Answering using LLM with Prompt Engineering

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [54]:
# Assigning the query to a variable
query1a = "What is the protocol for managing sepsis in a critical care unit?"
# Printing the response of query by changing the temperature parameter
print(response(query1a,temperature=0.5))

Llama.generate: prefix-match hit




Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following steps outline the general approach to managing sepsis in a critical care setting:

1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Look for signs of infection, such as fever or hypothermia, tachycardia or bradycardia, altered mental status, and lactic acidosis. Use validated scoring systems like the Sequential Organ Failure Assessment (SOFA) score or Quick Sequential Organ Failure Assessment (qSOFA) to help identify patients at risk for sepsis.
2. Resuscitation: Begin resuscitation efforts as soon as possible, focusing on maintaining adequate tissue perfusion and oxygenation. Administer intravenous fluids to maintain a mean arterial pressure of 65 mmHg or higher and to restore intravascular volume. Consider using vasopressors if blood pressure is n

It is observed that with change in temperature, the response became more creative and descriptive.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [55]:
# Assigning the query to a variable
query2a = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"

# Printing the response of query by changing the max_tokens parameter
print(response(query2a, max_tokens=1024))

Llama.generate: prefix-match hit




Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain may start as a mild discomfort around the navel or in the lower right abdomen, which then gradually moves to the right lower quadrant and becomes more severe over time. The pain may be constant or intermittent and is often worsened by movement, coughing, or deep breathing.
2. Loss of appetite: People with appendicitis may lose their appetite due to abdominal pain and discomfort.
3. Nausea and vomiting: Vomiting is a common symptom of appendicitis, especially in the later stages of the condition.
4. Fever: A fever of 100.4°F (38°C) or higher may be present in some cases of appendicitis.
5. Constipation or diarrhea: Both constipation and diarrhea can occur with appendicitis, depending on the location and severity

It is observed that upon increasing the max_tokens parameter, we are getting more descriptive reponse.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [56]:
# Assigning the query to a variable
query3a = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
# Printing the response of query by changing the top_p parameter
print(response(query3a,top_p=0.8))

Llama.generate: prefix-match hit




Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.

The exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.

There are several treatments that have been shown to be effective in addressing sudden patchy hair loss:

1. Corticosteroids: These are anti-inflammatory drugs that can help reduce inflammation and suppress the immune system's attack on the hair follicles. They can be applied topically or taken orally, depending on the severity of the condition.
2. Minoxidil: This is a medication that has been shown to promote hair growth in some people with alopecia areata. It works by increasing blood

With decrease in top_p parameter, we are getting more predictable but potentially incomplete response, suitable for fact retrieval but not for nuanced explanations.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [57]:
# Assigning the query to a variable
query4a = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
# Printing the response of query by changing the top_k parameter
print(response(query4a,top_k=30))

Llama.generate: prefix-match hit




A person who has sustained a physical injury to brain tissue, also known as a traumatic brain injury (TBI), may require various treatments depending on the severity and location of the injury. Here are some common treatments recommended for TBIs:

1. Emergency care: The first priority is to ensure the person's airway is clear, they are breathing, and their heart is beating normally. In severe cases, emergency surgery may be required to remove hematomas or other obstructions.
2. Medications: Depending on the symptoms, medications may be prescribed to manage conditions such as swelling, seizures, pain, or infections. For example, corticosteroids may be used to reduce brain swelling, and anticonvulsants may be given to prevent seizures.
3. Rehabilitation: Rehabilitation is an essential part of the recovery process for TBI patients. Rehabilitation may include physical therapy to help with mobility and strength, occupational therapy to help with daily living skills, speech therapy to impr

No major change is reponse is observed here after reducing top_k from 50 to 30.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [58]:
# Assigning the query to a variable
query5a = "List essential first-aid steps for a leg fracture during a hike, and key care & recovery guidelines after professional treatment."
# Printing the response of query by modifying the query
print(response(query5a))

Llama.generate: prefix-match hit




During a hike, if you or someone in your group experiences a leg fracture, follow these essential first-aid steps:

1. Safety: Ensure the safety of both the injured person and those around them. If necessary, move the person to a safe location away from any hazards such as traffic or steep terrain.
2. Call for help: Use a cell phone or communication device to call for emergency medical assistance if available. Inform them about the location, nature of the injury, and number of people involved.
3. Immobilize the leg: Use a splint, hiking pole, or any available materials to immobilize the fractured leg. This will help prevent further damage and provide comfort to the injured person.
4. Comfort: Keep the injured person warm and as comfortable as possible by covering them with a blanket or insulating material.
5. Monitor vital signs: Check for signs of shock, such as rapid heartbeat, shallow breathing, or pale skin. If these symptoms are present, try to keep the person calm and elevate t

We are getting a well structured and detailed response upon modifying the query.

## Data Preparation for RAG

### Loading the Data

In [6]:
# Import the Google Colab drive module
from google.colab import drive

# Mount Google Drive to the specified directory in the Colab environment.
# This will prompt you to authorize access.
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
# Path to the PDF file stored in your Google Drive
pdf_file = '/content/drive/My Drive/medical_diagnosis_manual.pdf'

In [8]:
# Create a PDF loader object to read and process the specified PDF file
pdf_loader = PyMuPDFLoader(pdf_file)

### Data Overview

#### Checking the first 5 pages

In [9]:
# Load all pages from the PDF
docs = pdf_loader.load()

# Display the first five pages (indexes 0–4)
for i in range(5):
    print(f"--- Page {i+1} ---")
    print(docs[i].page_content)   # or simply: print(docs[i])
    print("\n")

--- Page 1 ---
amit14rathor.in@gmail.com
W604NIQSDU
nt for personal use by amit14rathor.in@g
shing the contents in part or full is liable


--- Page 2 ---
amit14rathor.in@gmail.com
W604NIQSDU
This file is meant for personal use by amit14rathor.in@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.


--- Page 3 ---
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ..............................................................................................................................................................................

#### Checking the number of pages

In [10]:
# Check how many pages (documents) were loaded from the PDF
len(docs)

4114

There are 4114 pages in the pdf document.

### Data Chunking

In [11]:
# Create a text splitter that breaks the text into smaller chunks.
# Uses the tiktoken encoder for accurate token-based splitting.
# chunk_size = 512 tokens per chunk
# chunk_overlap = 16 tokens of overlap between consecutive chunks
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap=16
)

In [12]:
# Load the PDF and split it into smaller text chunks using the configured text splitter
document_chunks = pdf_loader.load_and_split(text_splitter)

In [13]:
# View the 7th text chunk (index 6) produced by the text splitter
document_chunks[6]

Document(metadata={'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'creationdate': '2012-06-15T05:44:40+00:00', 'source': '/content/drive/My Drive/medical_diagnosis_manual.pdf', 'file_path': '/content/drive/My Drive/medical_diagnosis_manual.pdf', 'total_pages': 4114, 'format': 'PDF 1.7', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'author': '', 'subject': '', 'keywords': '', 'moddate': '2025-11-30T12:33:09+00:00', 'trapped': '', 'modDate': 'D:20251130123309Z', 'creationDate': 'D:20120615054440Z', 'page': 4}, page_content='921\nChapter 94. Adrenal Disorders    ................................................................................................................................................\n936\nChapter 95. Polyglandular Deficiency Syndromes    ........................................................................................................\n939\nChapter 96. Porphyrias    ......................

In [16]:
# Installing required dependencies (using numpy<2 for compatibility with sentence-transformers)
!pip install -U "numpy<2" sentence-transformers langchain-huggingface

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 118.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 488.0/488.0 kB 43.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 475.0/475.0 kB 46.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 343.7/343.7 kB 36.5 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.3.3
    Uninstalling numpy-2.3.3:
      Successfully uninstalled numpy-2.3.3
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.80
    Uninstalling langchain-core-0.3.80:
      Successfully uninstalled langchain-core-0.3.80
  Attempting uninstall: sentence-transformers
    Found existing installation: sentence-transformers 5.1.1
    Uninstalling sentence-transformers-5.1.1:
      Successfully uninstalled sentence-transformers-5.1.1
ERROR: pip's dependency resolver does not currentl

### Embedding

In [14]:
# Initialize the embedding model that converts text chunks into numerical vectors.
# Using the 'thenlper/gte-large' model for high-quality semantic embeddings.
embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')

/tmp/ipython-input-2258052441.py:3: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the `langchain-huggingface package and should be used instead. To use it run `pip install -U `langchain-huggingface` and import as `from `langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')


modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

### Vector Database

In [15]:
# Name of the collection where embeddings will be stored
report = 'Article_QnA'

# Create a vector store using Chroma to store embeddings of the document chunks.
# - document_chunks: the text chunks to be embedded
# - embedding_model: the model used to generate embeddings
# - collection_name: the name of the collection in Chroma
vectorstore = Chroma.from_documents(
    document_chunks,
    embedding_model,
    collection_name=report
)


### Retriever

In [16]:
# Convert the Chroma vector store into a retriever object for semantic search
# - search_type='similarity': retrieves documents most similar to a query
# - search_kwargs={'k':2}: return the top 2 most relevant chunks for each query
retriever = vectorstore.as_retriever(
    search_type='similarity',
    search_kwargs={'k':2}
)


### System and User Prompt Template

In [17]:
# Defining  a controlled Q&A behavior where the model must answer only using the given context
qna_system_message = """
You are an assistant whose work is to give answers to questions with repect to a context.
User input will have the context required by you to answer user questions.

This context will begin with the token: ###Context.
The context contains references to specific portions of a document relevant to the user query.

User questions will begin with the token: ###Question.

Strictly answer only using the information provided in the ###Context.
Do not mention anything about the information in ###Context or the question in ###Question in your final answer.

If the answer to ###Question cannot be derived from the ###Context, just respond by saying "I don't know".

Remember that the answer to ###Question might not always be directly present in the information provided in the ###Context.
the answer can be indirectly derived from the information in ###Context.

"""

In [18]:
# This template formats the user message sent to the model.
# It inserts the context and the question under their respective tokens (###Context and ###Question),
# ensuring the model receives them in a structured and predictable format.
qna_user_message_template = """
Consider the following ###Context and ###Question
###Context
{context}

###Question
{question}
"""

### Response Function

In [19]:
import tiktoken

# Initialize tokenizer for counting and trimming tokens
enc = tiktoken.get_encoding("cl100k_base")

def trim_to_token_limit(text, max_tokens=1000):
    """
    Encodes the input text into tokens and trims it so that
    the total number of tokens does not exceed max_tokens.
    This prevents overly long contexts from breaking the model prompt.
    """
    tokens = enc.encode(text)
    if len(tokens) <= max_tokens:
        return text
    return enc.decode(tokens[:max_tokens])


def generate_rag_response(
        user_input,
        k=3,
        max_tokens=128,
        temperature=0,
        top_p=0.95,
        top_k=50
    ):
    global qna_system_message, qna_user_message_template

    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.invoke(user_input)
    context_list = [d.page_content for d in relevant_document_chunks[:k]]  # limit k

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    # Trim to safe token limit
    context_for_query = trim_to_token_limit(context_for_query, 1000)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + "\n" + user_message

    try:
        response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k
        )

        response = response["choices"][0]["text"].strip()
    except Exception as e:
        response = f"Sorry, I encountered the following error:\n{e}"

    return response


## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [20]:
# Assigning query to variable
Rquery1 = "What is the protocol for managing sepsis in a critical care unit?"
# Printing the prompt
print(generate_rag_response(Rquery1))

INFO:backoff:Backing off send_request(...) for 1.0s (requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')))


Answer: The protocol involves administering gentamicin or tobramycin along with a third-generation cephalosporin, ceftazidime, or ceftriaxone. If Pseudomonas is suspected, ceftazidime and a fluoroquinolone may be used instead. Vancomycin should be added if resistant staphylococci or enterococci are suspected. Metronidazole should be included if there is an abdominal source. Culture and sensitivity results should be used to adjust the antibiotic regimen accordingly.


Since this query was processed using a RAG setup with the supporting PDF document, the model’s response is expressed in precise medical terminology suitable for domain experts, which is exactly the expected behavior.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [21]:
# Assigning query to variable
Rquery2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
# Printing the prompt
print(generate_rag_response(Rquery2))

Llama.generate: prefix-match hit


I don't know if the context directly mentions the common symptoms for appendicitis or if it can be cured via medicine. However, I can tell you that appendicitis is a condition characterized by inflammation of the vermiform appendix and typically results in abdominal pain, anorexia, and abdominal tenderness. The treatment for appendicitis is surgical removal.


For this query, the symptoms and cure are directly not specified in the provided context, so the model correctly responds with “I don’t know” for the symptoms part. However, it indirectly gives the symptoms by stating “typically results in abdominal pain, anorexia, and abdominal tenderness.” For the rest of the question, the model provides accurate answers based on the available context.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [22]:
# Assigning query to variable
Rquery3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
# Printing the prompt
print(generate_rag_response(Rquery3))

Llama.generate: prefix-match hit


Based on the context provided, the possible causes of sudden patchy hair loss could be alopecia areata. The effective treatments for this condition include topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). It is important to note that the cause of sudden patchy hair loss cannot be definitively determined without a thorough evaluation. Therefore, it is recommended to consult


Both parts of the question, namely treatment and cause, are accurately explained by the model based on the information available in the PDF.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [23]:
# Assigning query to variable
Rquery4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
# Printing the prompt
print(generate_rag_response(Rquery4))

Llama.generate: prefix-match hit


Answer: The initial treatment consists of ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure. Surgery may be needed in patients with more severe injury to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important. Subsequently, many patients require rehabilitation. There is no specific treatment for the injury itself, but supportive care should include


The model is correctly retrieving the required answer for this query from the PDF context, providing appropriate recommendations for the treatment of a person suffering from brain tissue injury.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [24]:
# Assigning query to variable
Rquery5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# Printing the prompt
print(generate_rag_response(Rquery5))

Llama.generate: prefix-match hit


Answer:
The person with a fractured leg should elevate the injured limb above heart level for the first 2 days to minimize swelling. After 48 hours, they can apply warmth using a heating pad for pain relief and speeding up healing. Immobilization is necessary to prevent further injury and facilitate healing. Joints proximal and distal to the injury should be immobilized using either a cast or a splint. A cast should be kept dry, and patients should inspect the edges and skin around it daily for redness or soreness. They should also pad any rough edges with


The model is accurately explaining the precautions and treatment steps for a person with a fractured leg based on the information available in the PDF context.

## Inference:
When using RAG, the model’s responses are grounded in the specific context provided by the PDF, allowing it to give precise, accurate, and context-specific answers. Without RAG, the model generates responses based solely on its pre-trained knowledge, which may be less specific, potentially incomplete, or not aligned with the information in the PDF.

### Fine-tuning

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [25]:
# Assigning query to variable
Rquery1a = "What is the protocol for managing sepsis in a critical care unit?"
# Printing the prompt by changing temperature paramter
print(generate_rag_response(Rquery1a, temperature=0.5))

Llama.generate: prefix-match hit


Answer: The protocol involves administering gentamicin or tobramycin along with a third-generation cephalosporin, ceftazidime or ceftriaxone, or alternatively, ceftazidime and a fluoroquinolone for septic shock of unknown cause. Vancomycin must be added if resistant staphylococci or enterococci are suspected. If there is an abdominal source, a drug effective against anaerobes should be included. The antibiotic regimen is changed accordingly when culture and sensitivity results are


It is observed that upon increasing the temperature, the structure of response becomes more conversational and less exact but the meaning remains correct.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [39]:
# Assigning query to variable
Rquery2b = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
# Printing the prompt by changing the k paramter
print(generate_rag_response(Rquery2b, k=6))

Llama.generate: prefix-match hit


I don't know if the context directly mentions the common symptoms for appendicitis or if it can be cured via medicine. However, I can tell you that appendicitis is a condition characterized by inflammation of the vermiform appendix and typically results in abdominal pain, anorexia, and abdominal tenderness. The treatment for appendicitis is surgical removal.


Upon increasing k, the response remains same. May be all relevant info is already captured in the lower value of k that's why the output may remain unchanged.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [28]:
# Assigning query to variable
Rquery3c = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
# Printing the prompt by changing the top_k paramter
print(generate_rag_response(Rquery3c, top_k=30))

Llama.generate: prefix-match hit


Based on the context provided, the possible causes of sudden patchy hair loss could be alopecia areata. The effective treatments for this condition include topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). It is important to note that the cause of sudden patchy hair loss cannot be definitively determined without a thorough evaluation. Therefore, it is recommended to consult


No change in response observed after increasing top_k as the most relevant info is in the top few passages.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [59]:
# Assigning query to variable
Rquery4d = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
# Printing the prompt by changing the top_p paramter
print(generate_rag_response(Rquery4d,top_p=0.8))

Llama.generate: prefix-match hit


Answer: The initial treatment consists of ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure. Surgery is often needed in patients with more severe injury to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important. Subsequently, many patients require rehabilitation. There is no specific treatment for the brain injury itself, but supportive care should


Lowering top-p made the model more conservative and deterministic, producing slightly more precise but less varied response.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [30]:
# Assigning query to variable
Rquery5e = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# Printing the prompt by changing the max_tokens paramter
print(generate_rag_response(Rquery5e, max_tokens=1024))

Llama.generate: prefix-match hit


Answer:
The person with a fractured leg should elevate the injured limb above heart level for the first 2 days to minimize swelling. After 48 hours, they can apply warmth using a heating pad for pain relief and speeding up healing. Immobilization is necessary to prevent further injury and facilitate healing. Joints proximal and distal to the injury should be immobilized using either a cast or a splint. A cast should be kept dry, and patients should inspect the edges and skin around it daily for redness or soreness. They should also pad any rough edges with soft material to prevent discomfort. If an odor emanates from within the cast or a fever develops, medical care should be sought immediately. Good hygiene is important. Immobilization with bed rest can cause complications such as deep venous thrombosis and urinary tract infections. Prolonged immobilization (more than 3 to 4 weeks) can lead to stiffness, contractures, and muscle atrophy. For rapidly healing injuries, early mobilizatio

Increasing max_tokens allowed the model to generate a much longer, more detailed continuation, adding explanations about complications, hygiene, early mobilization, and care instructions that were previously cut off.

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [40]:
# This system message is designed for a groundedness rater.
# The model or evaluator will be given a question, the context used by the AI system,
# and the AI-generated answer. The task is to judge how much the answer is supported by the context.

groundedness_rater_system_message = """

You will be presented a ###Question, ###Context used by the AI system and AI generated ###Answer.

Your task is to judge the extent to which the ###Answer is derived from ###Context.

Rate it 1 - if The ###Answer is not derived from the ###Context at all
Rate it 2 - if The ###Answer is derived from the ###Context only to a limited extent
Rate it 3 - if The ###Answer is derived from ###Context to a good extent
Rate it 4 - if The ###Answer is derived from ###Context mostly
Rate it 5 - if The ###Answer is is derived from ###Context completely

Please note: Make sure you give a single overall rating in the range of 1 to 5 along with an overall explanation.

"""

In [41]:
# This system message is designed for a relevance rater.
# The model or evaluator will be given a question, the context used by the AI system,
# and the AI-generated answer. The task is to judge how relevant the answer is to the question.

relevance_rater_system_message = """

You will be presented with a ###Question, the ###Context used by the AI system to generate a response, and the AI-generated ###Answer.

Your task is to judge the extent to which the ###Answer is relevant to the ###Question, considering whether it directly addresses the key aspects of the ###Question based on the provided ###Context.

Rate the relevance as follows:
- Rate 1 – The ###Answer is not relevant to the ###Question at all.
- Rate 2 – The ###Answer is only slightly relevant to the **###Question**, missing key aspects.
- Rate 3 – The ###Answer is moderately relevant, addressing some parts of the **###Question** but leaving out important details.
- Rate 4 – The ###Answer is mostly relevant, covering key aspects but with minor gaps.
- Rate 5 – The ###Answer is fully relevant, directly answering all important aspects of the **###Question** with appropriate details from the **###Context**.

Note: Provide a single overall rating in the range of 1 to 5, along with a brief explanation of why you assigned that score.
"""

In [42]:
# This template formats the input for evaluation by a rater (groundedness or relevance).
# It structures the message by placing the question, context, and AI-generated answer
# under clearly labeled sections (###Question, ###Context, ###Answer).
# The placeholders {question}, {context}, and {answer} are replaced with the actual values
# before sending to the model or evaluator.

user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [43]:
# This function generates a response from the model specifically for evaluating
# groundedness or relevance of an AI-generated answer.
def generate_ground_relevance_response(user_input, k=3, max_tokens=128, temperature=0, top_p=0.95, top_k=50):
    global qna_system_message, qna_user_message_template

    # Retrieve relevant document chunks - USING invoke()
    relevant_document_chunks = retriever.invoke(user_input)  # Changed here
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer = response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'], response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [44]:
# Assigning variable to the query
Equery1 = "What is the protocol for managing sepsis in a critical care unit?"
# Printing the response
print(generate_ground_relevance_response(Equery1))

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


(' Rating: 5 - The Answer is derived from the Context completely.\n\nExplanation: The context provides a detailed description of the approach to managing sepsis in a critical care unit, including the importance of prompt empiric antibiotic therapy and the specific regimens mentioned in the answer. Therefore, the answer is a direct reflection of the information provided in the context.', ' Rating: 5\n\nExplanation: The answer directly addresses the key aspects of managing sepsis in a critical care unit as outlined in the question, including giving parenteral antibiotics after taking specimens for Gram stain and culture, starting empiric therapy immediately with appropriate antibiotics based on suspected source and organisms, adding Vancomycin if resistant staphylococci or enterococci are suspected, and adjusting the antibiotic regimen based on culture and sensitivity results. The answer also mentions continuing antibiotics for at least 5 days after shock res')


Both explanations give a rating of 5, agreeing that the answer is fully derived from the context. They note that the answer accurately reflects all key steps in managing sepsis in a critical care unit—such as prompt empiric antibiotics, culture-guided adjustments, and specific drug options—exactly as described in the provided context.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [45]:
# Assigning variable to the query
Equery2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
# Printing the response
print(generate_ground_relevance_response(Equery2))

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


(' Rating: 5 - The Answer is derived from the Context completely.\n\nExplanation: The Context clearly states that the treatment for appendicitis is surgical removal, specifically an open or laparoscopic appendectomy. It also mentions that antibiotics are used prior to surgery and in some cases after surgery if the appendix has perforated. The Answer accurately reflects this information by stating that common symptoms include abdominal pain, anorexia, and abdominal tenderness, and that surgical removal is required for treatment as appendicitis cannot be cured via medicine alone.', ' Rating: 5\n\nExplanation: The AI-generated answer directly addresses both parts of the question by correctly identifying the common symptoms of appendicitis and stating that it cannot be cured with medicine, but rather requires surgical removal. This information is fully relevant to the question based on the provided context.')


Both explanations agree that the response fully reflects the context, correctly identifying symptoms and surgical treatment, and they both give a rating of 5 because the answer is accurate, complete, and directly supported by the context.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [46]:
# Assigning variable to the query
Equery3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
# Printing the response
print(generate_ground_relevance_response(Equery3))

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


(" Rating: 5 - The Answer is derived from the Context completely.\n\nExplanation: The user's question asks about effective treatments and possible causes for sudden patchy hair loss, which is a description of alopecia areata as presented in the context. The answer accurately identifies alopecia areata as an autoimmune disorder and lists the various treatment options mentioned in the context, including topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squar", ' Rating: 5\n\nExplanation: The question asks about the effective treatments and causes of sudden patchy hair loss, which is exactly what the context provides information on in relation to alopecia areata. The answer directly addresses the key aspects of the question by listing the various treatment options for alopecia areata and mentioning that it is an autoimmune disorder.')


Both explanations agree that the response fully addresses the question, correctly identifying alopecia areata as an autoimmune disorder and listing all relevant treatments, resulting in a rating of 5 for being complete and contextually accurate.

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [47]:
# Assigning variable to the query
Equery4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
# Printing the response
print(generate_ground_relevance_response(Equery4))

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


(' Rating: 5 - The Answer is derived from the Context completely.\n\nExplanation: The context provides detailed information about the initial treatment and supportive care for a person with traumatic brain injury (TBI). The Answer accurately summarizes these details, including ensuring a reliable airway, maintaining adequate ventilation, oxygenation, and blood pressure, surgery to place monitors or decompress the brain if necessary, preventing complications in the first few days, and providing supportive care such as preventing systemic complications due to immobilization and providing good nutrition.', ' Rating: 5 - The Answer is fully relevant, directly answering all important aspects of the Question with appropriate details from the Context.\n\nExplanation: The Question asks about treatments for a person with brain injury resulting in impairment of brain function. The Context provides information on the initial treatment for such patients which includes ensuring a reliable airway an

Both explanations agree that the response fully and accurately reflects the context, covering initial treatment, surgical interventions, early supportive care, and ongoing management for traumatic brain injury, resulting in a rating of 5 for completeness and relevance.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [48]:
# Assigning variable to the query
Equery5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# Printing the response
print(generate_ground_relevance_response(Equery5))

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


(' I would rate the answer as a 4 out of 5 in terms of being derived from the context. The answer accurately summarizes the key points from the context regarding the necessary precautions and treatment steps for someone with a fractured leg, including the use of ice, immobilization using a splint or cast, elevation of the injured limb, instructions for patients with casts, and the potential benefits and risks of prolonged immobilization. The answer also includes some additional information about complications that may arise from fractures, but this information is not explicitly stated in the context. Overall, the answer provides a', ' Rating: 5\n\nExplanation: The answer directly addresses the key aspects of the question by summarizing the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. It mentions the application of ice, immobilization using a splint or cast, elevation of the injured limb above heart level, and instructions for 

The first explanation rates the answer 4/5 because it includes some extra information not explicitly in the context, while the second rates it 5/5 since it views all details—including early mobilization and cast care—as fully supported by the context. The discrepancy arises from whether additional inferred details are considered outside the context.

## Actionable Insights

1. The AI system performs excellently (5-star ratings) on standard clinical queries but shows a slight drop (4-star rating) for complex, multi-part, situational queries like the hiking fracture scenario.

2. The system excels at factual retrieval but struggles slightly with multi-phase, scenario-based questions requiring synthesized guidance.

3. The system is production-ready for standard clinical use, with clear path to dominate complex scenario markets with targeted enhancements.

## Business Recommendations

1. Healthcare centers can deploy the model in their facilities during the production phase for emergency use.

2. Centers can display confidence levels per answer section and allow natural follow-up questions so users can drill deeper into areas like post-fracture recovery or specific precautions.

3. Maximum usage in urgent care centers and remote-consultation platforms will be more helpful where complex, situational questions are common, using their feedback to refine scenario-based responses before wider rollout.